OpenAPI Export for API Definition, Swagger UI Binding

Nanashi-lab opened the pull request:
**PR: OpenAPI Export and Swagger UI Support for Golem API Definitions**  Closes #1178 
/claim #1178 

This PR provides OpenAPI export functionality and Swagger UI (Binding and CLI) support. Specifically, this PR implements:

*   **Key Features:**
    *   [x] **OpenAPI Export Endpoint:** Added `/v1/api-definitions/{id}/{version}/export` endpoint to export API Definitions to OpenAPI Specifications.
    *   [x] **Full and Lossless Conversion:** Implemented  conversion logic for request and response types,  accurate mapping of type to OpenAPI schema.
    *   [x] **Swagger UI Binding Type:** A new "Swagger UI" binding type for routes in the Worker Gateway, allowing for serving of Swagger UI interfaces from API definitions when deployed.
    *   [x] **Swagger UI Integration:** Code to render the exported OpenAPI schema as a Swagger UI for user-friendly API testing.
    *   [x] **Testing:** Added basic unit for API to OpenAPI spec, tests include, verifying CORS, secuirty, parameter, input and output maping, I have also added end-to-end tests to validate Swagger UI.
    *   [x] **Golem CLI Updates:** `golem-cli` needs to be updated with:
        *   [x] `golem-cli api-definition swagger` subcommand for launching a browser to a generated Swagger UI.
        *   [x] `golem-cli api-definition export` subcommand to export an API definition to an OpenAPI schema.
    *   [ ] **Documentation:** Documentation is required to explain these new features and will be provided in a separate PR to the docs repository.
    *   [x] **System/Integration Tests:** 
        *   [x] System tests to OpenAPI Spec from API, and attempting full circle by importing the OpenApi Schema,

System test and verify they interact correctly with the SWAGGER UI (Might not be possible currently all deployment tests fail.)

You can try this out, after building, by creating a new route with `bindingtype swagger-ui`, with path without any varriables, example `/swagger` . If deployed to localhost:9006, you will find your swagger ui there.

Your feedback is appreciated! While I work on the remaining task. 

![Capture](https://github.com/user-attachments/assets/a3cc471c-48df-4c5f-a746-781f80881f3d)
![Capture2](https://github.com/user-attachments/assets/dd2a8ed1-f036-4578-aa65-ca916f0a4404)

Nanashi-lab commented on golem-worker-service-base/src/gateway_execution/gateway_http_input_executor.rs:
```diff
@@ -801,36 +822,6 @@ fn to_attribute_value(value: &ValueAndType) -> GatewayHttpResult<AttributeValue>
-        ApiDefinitionLookupError::UnknownSite(_) => StatusCode::NOT_FOUND,
-    }
-}
-
```
This wasn't meant to be removed, I will fix it with a commit.

Nanashi-lab replied:
Fixed this

Nanashi-lab commented:
PR for CLI Part of the task is [here](https://github.com/golemcloud/golem-cli/pull/159) 
There are two new command under `golem-cli api definition`
`swagger` and `export`

`golem-cli api definition export --id <id> --version <version>  ` will download the open api spec in either json or yaml format, this will be the whole open api schema including input and output type 


![export](https://github.com/user-attachments/assets/25338c91-a8f5-4f0a-8f4d-0c44cacc4d35)

`golem-cli api definition export --id <id> --version <version> --host <host>` will open the webbrowser and navigate to the host (usually localhost:<port>), if the api has deployments it will add the deployments in server, to test. else it will show the open api without any server information 


![swagger](https://github.com/user-attachments/assets/3632429a-dc4d-414d-ae9f-a2abea1890dd)

example of the swagger ui usage in localhost:9900, it rightly shows 9006 (local deployment) as server for testing


![Capture3](https://github.com/user-attachments/assets/0b1c0898-9ded-48b0-93e0-e57fe375ac5c)

jdegoes commented on golem-worker-service-base/src/gateway_api_definition/http/api_oas_convert.rs:
```diff
@@ -0,0 +1,1197 @@
+            {
+                // Simple heuristic to detect if the response might have multiple status codes
+                // TODO: This is a simple heuristic and might not be 100% accurate
+                response.contains("status:")
```
Rib can give you this information through type inference (OneOf constructor). There is no need to resort to string-scanning.

Nanashi-lab replied:
That is a much simpler, and better solution, than string scanning. Done.

Nanashi-lab replied:
I am having trouble passing the right code. example, post method expects a response at 201, but rib script can send a response which is 200
similarly if {status:u32, string} is the type output, it is possible to pass {200, "success"}, {400, "failure"}

solution
have a response code from the method, post 201 get 200 etc
have an another response:default -> which also uses the same type data as the specific code. 

Swagger UI deals with this much better.

jdegoes commented on golem-worker-service-base/src/gateway_api_definition/http/api_oas_convert.rs:
```diff
@@ -0,0 +1,1197 @@
+                if !response_mapping_input.types.is_empty() {
+                    // Check if the response uses request.body
+                    let uses_request_body = if let Some(response) = &route.binding.response {
+                        response.contains("request.body")
```
These string-scans need to be avoided because they could pick up on a comment, e.g. `// todo: use request.body` and generate false positives.

Nanashi-lab replied:
I have removed string scanning, and the code rely only on input mapping, and output mapping to determine if we have requestBody and responses

jdegoes commented on golem-worker-service-base/src/gateway_api_definition/http/api_oas_convert.rs:
```diff
@@ -0,0 +1,1197 @@
+
+                    let mut properties = indexmap::IndexMap::new();
+                    properties.insert(
+                        case_name.clone(),
```
Testing

jdegoes commented on golem-worker-service-base/tests/api_oas_convert_tests.rs:
```diff
@@ -0,0 +1,960 @@
+    let yaml_str = openapi_response.openapi_yaml;
+    assert!(yaml_str.contains("component-name: shopping-cart"));
+    assert!(yaml_str.contains("component-version: 0"));
+    assert!(yaml_str.contains("let item = request.body.item; let result = golem:shoppingcart/api.{add-item}(item); {status: 200u64, body: \"Item added\"}"));
```
These are junk tests not actually verifying that the OpenAPI schema correctly represents the input and output of the endpoint (expressed in JSON Schema).

This test is just verifying the original Rib code can be found somewhere in the YAML which is useless.

Nanashi-lab replied:
I have improved the test to now test for structure rather than just if string contains.
   
   most of the tests now test for actual structure
   
    let responses = &yaml_value["paths"]["/v0.1.0/api/resource"]["options"]["responses"];
    assert!(responses["200"].is_mapping());
    assert_eq!(responses["200"]["description"], "OK");
    
    I have also added one test, which tests the whole openapi schema against expected schema for easy readability

jdegoes commented on golem-worker-service-base/tests/api_oas_convert_tests.rs:
```diff
@@ -0,0 +1,960 @@
+
+    // Check CORS binding and configuration
+    assert!(yaml_str.contains("binding-type: cors-preflight"));
+    assert!(yaml_str.contains("Access-Control-Allow-Headers: \"Content-Type, Authorization\""));
```
Junk test.

jdegoes commented on golem-worker-service-base/tests/api_oas_convert_tests.rs:
```diff
@@ -0,0 +1,960 @@
+    assert!(yaml_str.contains("requestBody:"));
+    assert!(yaml_str.contains("required: true"));
+    assert!(yaml_str.contains("application/json:"));
+    assert!(yaml_str.contains("type: object"));
```
Junk test.

You should explicitly pull out the type schema for request and response and verify that it's as expected (you could do JSON normalization to make testing easier).

jdegoes commented on golem-worker-service-base/tests/api_oas_convert_tests.rs:
```diff
@@ -0,0 +1,960 @@
+    assert!(yaml_str.contains("properties:"));
+    assert!(yaml_str.contains("body:"));
+    assert!(yaml_str.contains("type: array"));
+    assert!(yaml_str.contains("items:"));
```
I won't comment anymore but all these tests are junk tests.

Nanashi-lab replied:
I understand, I will remove the junk tests, add fewer, but more comprehensive tests, which tests the whole responseOuput, responseInput structure, than just verifying that the yaml contains strings.

Nanashi-lab replied:
Finally, I have improved the code readability of api_oas_convert, by dividing it into smaller easy to read functions, with clear explanations.

Nanashi-lab commented:
This is a link to [repository](https://github.com/Nanashi-lab/api-oas-convert) containing, some wasm-files, api json files and the equivalent yaml output files.

Nanashi-lab commented:
All Checks have passed, that's really good news.

Nanashi-lab commented on golem-worker-service-base/tests/api_gateway_end_to_end_tests.rs:
```diff
@@ -3634,3 +3663,114 @@ nUhg4edJVHjqxYyoQT+YSPLlHl6AkLZt9/n1NJ+bft0=
+}
+
+#[test]
+async fn test_swagger_ui_integration() {
```
This just the repeat test of above, what was meant to be here, is a test, which does a mock usage of the swagger ui, interacting with mock worker, and testing that only certain bindings are show in swagger ui, properly. My fault. I will fix this.

Nanashi-lab commented:
I matched HEAD with current golem, 
Few other changes are 
1) Removed extract_path_parameters, we get path details directly input_mapping
2) Added test for first class worker 
3) Added test for variant and query
4) Cleaned up the repeated test.

Nanashi-lab commented:
One of my tests failed 

```rust
ERROR test_export_import_api_definition{deps=EnvBasedTestDependencies}: golem_common::tracing: panic panic_info=panicked at integration-tests/tests/api/api_definition.rs:587:10:
called `Result::unwrap()` on an `Err` value: ApiDefinitionError { error: Some(BadRequest(ErrorsBody { errors: ["Invalid JSON"] })) }"
```

I will figure out why this happening.

Nanashi-lab commented:
@jdegoes @vigoo Gentle Reminder - 

Please review & offer feedback on this PR. To ensure it's as easy as possible to review, I've rebased  both `golem` and `golem-cli`  with the latest main branch, so it is up-to-date. 

Could you also let me know the current state of the ticket? Is the PR/Issue still something the team is actively interested in? Knowing this will help me prioritize accordingly. 

I want to ensure I'm providing everything you need to make a decision on this PR. Please let me know what I can do to help move it forward. 

To make this process smoother, I'm happy to:

- Set up a live demo to walk you through the changes.
- Provide any additional technical information or documentation you need.
- Answer any questions you may have.

afsalthaj commented:
@Nanashi-lab  I am currently taking a look at this PR

afsalthaj commented on golem-test-framework/src/components/worker_service/mod.rs:
```diff
@@ -1120,6 +1124,34 @@ pub trait WorkerService: WorkerServiceInternal {
         }
     }
 
+    async fn export_api_definition(
```
Is `export_api_definition` the name we chose?
If this is about mainly creating open api spec, then can we have that in the naming across?

Nanashi-lab replied:
Done, Renamed to `export_openapi_spec` for readability, across all tests and implementation.

afsalthaj commented on golem-worker-service-base/src/api/api_definition.rs:
```diff
@@ -310,8 +310,8 @@ pub struct GatewayBindingComponent {
-    name: String,
-    version: u64,
+    pub name: String,
+    pub version: u64,
```
Let's don't make this public, probably this is constructed in specific places.
and we want to guarantee it shouldn't be called from any random places.

afsalthaj replied:
You can just some functions and access name and version

afsalthaj replied:
Also I was wondering why this is changing or used in the open api export.

Nanashi-lab replied:
Done
```rust
// Getters for Component Name and Version, Used in OpenAPI Spec conversion
impl ResolvedGatewayBindingComponent {
    pub fn name(&self) -> &str {
        &self.name
    }
    pub fn version(&self) -> u64 {
        self.version
    }
}
```

We use this to get name and version of component in OpenAPI

```yaml
      x-golem-api-gateway-binding:
        binding-type: default
        component-name: shopping-cart
        component-version: 0
```

Nanashi-lab commented on golem-common/src/model/mod.rs:
```diff
@@ -2122,7 +2124,10 @@ impl TryFrom<String> for GatewayBindingType {
+            "http-handler" => Ok(GatewayBindingType::HttpHandler),
+            "cors-preflight" => Ok(GatewayBindingType::CorsPreflight),
+            "swagger-ui" => Ok(GatewayBindingType::SwaggerUi),
+            _ => Err(format!("Invalid GatewayBindingType: {}", value)),
```
> I am also not sure why we need this TryFrom which is a source of bug. I am pretty sure it's unused, that everything works even when there was no http-handler, and cors-prelight

This is my fault, After implementing the swagger binding type, I added it without much for thought when adding swagger-ui binding type, to other parts of golem.

Since wit-worker compatibility is not required, 
Removed both the curstom deserialize, and try from, added `derive Deserialize`

afsalthaj commented on golem-worker-service-base/src/gateway_api_definition/http/api_oas_convert.rs:
```diff
@@ -0,0 +1,1097 @@
+// Convert HttpApiDefinitionResponseData to OpenHttpApiDefinition
+impl OpenApiHttpApiDefinition {
+    pub fn from_http_api_definition_response_data(
+        response_data: &HttpApiDefinitionResponseData,
```
Can this function take `CompiledHttpApiDefinition<Namespace>,` as input instead of `HttpApiDefinitionResponseData` ?


`HttpApiDefinitionResponseData` is a subset of  `CompiledHttpApiDefinition<Namespace>,`  and may not have all the information. 

More importantly, `CompiledHttpApiDefinition<Namespace>` is typesafe that it gives us more confidence that openAPI spec is correct.


```rust
// part of safe CompiledHttpApiDefinition
#[derive(Debug, Clone, PartialEq)]
pub enum GatewayBindingCompiled {
    Worker(WorkerBindingCompiled),
    Static(StaticBinding),
    FileServer(WorkerBindingCompiled),
    HttpHandler(HttpHandlerBindingCompiled),
    SwaggerUi,
}
```

vs

```rust
// part of unsafe HttpApiDefinitionResponseData
pub struct GatewayBindingResponseData {
    pub component: Option<ResolvedGatewayBindingComponent>, // Optional to keep it backward compatible
    pub worker_name: Option<String>, // If bindingType is Default or FileServer
    pub idempotency_key: Option<String>, // If bindingType is Default or FileServer
    pub response: Option<String>, // Optional to keep it backward compatible. If bindingType is Default or FileServer
    #[oai(rename = "bindingType")]
    pub binding_type: Option<GatewayBindingType>,
    pub response_mapping_input: Option<RibInputTypeInfo>, // If bindingType is Default or FileServer
    pub worker_name_input: Option<RibInputTypeInfo>,      // If bindingType is Default or FileServer
    pub idempotency_key_input: Option<RibInputTypeInfo>, // If bindingType is Default or FilerServer
    pub cors_preflight: Option<HttpCors>, // If bindingType is CorsPreflight (internally, a static binding)
    pub response_mapping_output: Option<RibOutputTypeInfo>, // If bindingType is Default or FileServer
}
```

Nanashi-lab replied:
Yes, I will Figure this out CompiledHttpApiDefinition directly to ExportOpenapiSpec. I am unsure how the test will look, as now we just convert between JSON representation of HttpApiDefinitionResponseData and convert it into ExportOpenapiSpec yaml, I will look into this. Also refrain from editing/adding commit to the code during the review, sorry about that.

afsalthaj commented on golem-worker-service-base/src/gateway_api_definition/http/api_oas_convert.rs:
```diff
@@ -0,0 +1,1097 @@
+        openapiv3::ReferenceOr<openapiv3::SecurityScheme>,
+    >,
+) -> Result<(), String> {
+    let path_str = route.path.to_string();
```
isn't `route.path` already a String?

afsalthaj commented on golem-worker-service-base/src/gateway_api_definition/http/api_oas_convert.rs:
```diff
@@ -0,0 +1,1097 @@
+
+        // Handle integer types using the macro for the body
+        AnalysedType::U8(_) => {
+            create_integer_schema_body!(openapiv3::IntegerFormat::Int32, Some(0), Some(255))
```
Minor: So avoid macros I guess :)

afsalthaj commented:
@Nanashi-lab  I would request you to commit the fixes after I finish the review - I will let you know when, otherwise hard for me to review :D 

It's looking good already, but just need a couple of more hours.

afsalthaj commented on golem-worker-service-base/src/gateway_api_definition/http/api_oas_convert.rs:
```diff
@@ -0,0 +1,1097 @@
+        400 => "Bad Request".to_string(),
+        404 => "Not Found".to_string(),
+        500 => "Internal Server Error".to_string(),
+        _ => "Response".to_string(),
```
_ = "Response" ?
Is it the standard pattern?

afsalthaj commented on golem-worker-service-base/src/gateway_api_definition/http/api_oas_convert.rs:
```diff
@@ -0,0 +1,1097 @@
+}
+
+// Helper function: Determines response schema
+fn determine_response_schema(route: &RouteResponseData) -> openapiv3::Schema {
```
I think better to use `CompiledRoute` instead of `RouteResponseData`./
I think you will end up having to use it because of [this](https://github.com/golemcloud/golem/pull/1454/files#r2093467076)

afsalthaj commented on golem-worker-service-base/src/gateway_api_definition/http/api_oas_convert.rs:
```diff
@@ -0,0 +1,1097 @@
+    if let Some(response_mapping_output) = &route.binding.response_mapping_output {
+        if let AnalysedType::Record(record) = &response_mapping_output.analysed_type {
+            let has_status = record.fields.iter().any(|f| f.name == "status");
+            let has_body = record.fields.iter().any(|f| f.name == "body");
```
In the response there can be headers too. So we need the schema for response headers.
In the request too, there are headers. I guess we are not looking up it's schema as well.

https://learn.golem.cloud/rib#external-interactions

afsalthaj commented on golem-worker-service-base/src/gateway_api_definition/http/api_oas_convert.rs:
```diff
@@ -0,0 +1,1097 @@
+                    .find(|f| f.name == "body")
+                    .map(|body_field| create_schema_from_analysed_type(&body_field.typ))
+                    .unwrap_or_else(create_default_object_schema),
+                (true, false) => create_default_object_schema(),
```
I would rather refactor this to the following
If it's a `AnalysedType::Record`, then there are chances that it has a body key, status key and headers key. Instead of permuting through all the combinations of their existence or absence, simply introduce three functions that returns an Option

```rust
fn extract_header_type(record_name_type_pairs: &[NameTypePair]) -> Option<Vec<(HeaderNameString, AnalysedType)>>;

fn extract_body_type(record_name_type_pairs: &[NameTypePair]) -> Option<AnalysedType>;

fn extract_status() 
fn extract_response_content_type() -> Option<String>

```

```
let headers_opt = extract_...
let body_opt = extract...
let status_opt = extract...

if headers._opt.is_some() | body_opt.is_some() | status_opt.is_some()  {
       let body = body_opt.unwrap_or_else(|| create_default_object_schema);
       let headers = header_opt.unwrap_or_else(...default schema)
       let status = status_opt.unwrap_or_else(.. default schema)
} else {
  create_schema_from_analysed_type(&response_mapping_output.analysed_type)
}

```

Something along those lines

afsalthaj replied:
Something along those lines makes it easy to understand, avoid potential bugs..
Even verbose yet bug-free implementation would be to do pattern match, such that tomorrow if we have any additions, compiler will ask to fix instead of being happy with "headers._opt.is_some() | body_opt.is_some() | status_opt.is_some()"

afsalthaj commented on golem-worker-service-base/src/gateway_api_definition/http/api_oas_convert.rs:
```diff
@@ -0,0 +1,1097 @@
+        Some(GatewayBindingType::HttpHandler) => {
+            binding_info.insert(
+                "binding-type".to_string(),
+                serde_json::Value::String("http-handler".to_string()),
```
Good to reus the serialize deserialize of the binding type used in API calls

afsalthaj commented on golem-worker-service-base/src/gateway_api_definition/http/api_oas_convert.rs:
```diff
@@ -0,0 +1,1097 @@
+
+// Helper function: Adds cors-preflight binding info
+// Converts the cors-preflight binding info to a response string
+fn add_cors_preflight_binding_info(
```
A binding is per route.  Per route, there cannot exist multiple binding info
This is more tangible if you use CompiledRoute instead of the `RouteResponseData` which is flattened structure making all binding optional

afsalthaj commented on golem-worker-service-base/src/gateway_execution/swagger_binding_handler.rs:
```diff
@@ -0,0 +1,324 @@
+
+impl<Namespace> DefaultSwaggerBindingHandler<Namespace> {
+    pub fn new(
+        api_definition_lookup_service: Arc<dyn HttpApiDefinitionsLookup<Namespace> + Sync + Send>,
```
It looks a bit off to lookup API definition again, if the binding is swagger.

So if I understand, what's implemented is as follows:
* When the user hits the swagger route, gateway looks up the set of api definition based on the authority.(part of standard worfklow of gateway)
* It then identify the real api definition that has the same route (swagger route in this case) and identifies its a swagger binding.(part of standard workflow)
* Then it goes to swagger-binding-handler, where  it looks up _again_ (concerning lookup this is) to get the set of api definitions corresponding to the authority 
* Then it filters down to the real one that has a swagger binding with the same route path.

AFAIU the API definition that we retrieved as part of the last step is the same as the one we identified in the second step. 

I understand why this approach was taken because the API definition retrieved in the second step is probably not quite visible from a coding perspective. While this is easier approach but it is a confusing workflow.

A better approach therefore is CompiledHttpAPIdefinition can store the open API spec itself when it is swagger, such that the following:

```rust
GatewayBindingCompiled::SwaggerUi => {
                let result = self
                    .swagger_binding_handler
                    .handle_swagger_binding_request(&namespace, &authority, &rich_request)
                    .await;

```
will be replaced by

```rust
GatewayBindingCompiled::SwaggerUi(openAPIspec)  => {
                let result = self
                    .swagger_binding_handler
                    .handle_swagger_binding_request(&namespace, open_api_spec)
                    .await;

````

afsalthaj replied:
This will reduce the complexity such as ""Definition service not available" or "   "Swagger UI cannot have variable path".to_string()" etc in the code - I guess!

afsalthaj commented on golem-worker-service-base/tests/api_oas_convert_tests.rs:
```diff
@@ -0,0 +1,1640 @@
+                    },
+                    "response": "let item = request.body.item; let result = golem:shoppingcart/api.{add-item}(item); {status: 200u64, body: \"Item added\"}",
+                    "workerName": "let user: string = request.path.user; \"worker-${user}\"",
+                    "workerNameInput": {
```
I guess `workerNameInput` shouldn't exist in any tests.

Let's have simple yaml API definitions (see api_gateway_end_to_end_tests) and simply a swagger ui binding.

The API definitions can have 10-15 routes covering many possibilities of input and output type (meaning 15 types of rib scripts). 

And hand written expectation of the open-api spec, corresponding to input api definition. This is what I would personally write as tests.

afsalthaj reviewed (changes requested):
I have made some comments. There are a few things to fix before we could merge.  I can see you are getting into various internals to reach this point, and pretty sure we can make it all the way if you could fix these comments soon :)

jdegoes commented:
@Nanashi-lab Please re-open when 100% of all feedback has been addressed in the way indicated and you are ready for a fresh review. Thank you!
